File: dependencies/SIFT-private/external/tsa/minimal/sumskipnan.cc
#include <octave/oct.h>
DEFUN_DLD (sumskipnan, args, nargout, "OCT-implementation of SUMSKIPNAN\n") 
//   OCT implementation of SUMSKIPNAN - this function is part of the NaN-toolbox. 
//
//   This program is free software; you can redistribute it and/or modify
//   it under the terms of the GNU General Public License as published by
//   the Free Software Foundation; either version 2 of the License, or
//   (at your option) any later version.
//
//   This program is distributed in the hope that it will be useful,
//   but WITHOUT ANY WARRANTY; without even the implied warranty of
//   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
//   GNU General Public License for more details.
//
//   You should have received a copy of the GNU General Public License
//   along with this program; if not, write to the Free Software
//   Foundation, Inc., 59 Temple Place, Suite 330, Boston, MA  02111-1307  USA
//
//
 	// sumskipnan: adds all non-NaN values
//
// Input:
// - array to sum
// - dimension to sum (1=columns; 2=rows; doesn't work for dim>2!!)
//
// Output:
// - sums
// - count of valid elements (optional)
// - sums of squares (optional)
// - sums of squares of squares (optional)
//
// Y = sumskipnan(x [,DIM])
// [Y,N,SSQ] = sumskipnan(x [,DIM])
// 
// DIM	dimension
//	1 sum of columns
//	2 sum of rows
//	default or []: first DIMENSION with more than 1 element
//
// Y	resulting sum
// N	number of valid (not missing) elements
// SSQ	sum of squares
//
//	$Id: sumskipnan.cc,v 1.3 2005/04/28 21:03:28 schloegl Exp $
//    Copyright (C) 2005 by Alois Schloegl <a.schloegl@ieee.org>	
//    This is part of the NaN-toolbox 
//    http://www.dpmi.tu-graz.ac.at/~schloegl/matlab/NaN/


{ 
  int DIM = 0;  
  int ND;
  int j, k1, k2, k3, D1, D2, D3, ix1, ix2, N;
  dim_vector SZ, SZ2;
  double aux, s, si; 
  double n, s2, s4;
  bool IS_COMPLEX;

  octave_value_list retval; 

  int nargin = args.length ();

  if ( args.length() == 2 ) 
  {
	  Matrix xin2( args(1).matrix_value() );
	  DIM = (int) xin2(0);
  }
  else if ( args.length() > 3 ) 
  {
    error("SUMSKIPNAN.OCT takes at most 2 arguments");
    return retval;
  }

	// only real double matrices are supported; arrays, integer or complex matrices are not supported (yet)
	IS_COMPLEX = args(0).is_complex_type ();
	ND = args(0).ndims ();    
	SZ = args(0).dims ();

    	for (j = 0; (DIM < 1) && (j < ND); j++) 
		if (SZ(j)>1) DIM = j+1;

  	if (DIM < 1) DIM=1;		// in case DIM is still undefined 
	
	SZ2 = SZ; 
	//for (j=ND; j<DIM; SZ2(j++)=1);
	if (DIM>ND)
	{    	error("SUMSKIPNAN.OCT: DIM larger than LENGTH(SIZE(X))");
		return retval;
	}	
	if (IS_COMPLEX)
	{    	warning("SUMSKIPNAN.OCT: imaginary part ignored.");
	}	
			
   	for (j=0, D1=1; j<DIM-1; D1=D1*SZ(j++)); 	// D1 is the number of elements between two elements along dimension  DIM  
	D2 = SZ(DIM-1);					// D2 contains the size along dimension DIM 	
    	for (j=DIM, D3=1;  j<ND; D3=D3*SZ(j++)); 	// D3 is the number of blocks containing D1*D2 elements 

  	SZ2(DIM-1) = 1;

	// Currently, only real arrays are supported. In future, also complex arrays should be supported. 
 	NDArray xin1(args(0).array_value ());
 	NDArray xout1(SZ2);
//	NDArray xout1i(SZ2);
	NDArray xout2(SZ2);
	NDArray xout3(SZ2);
	NDArray xout4(SZ2);
	
	// OUTER LOOP: along dimensions > DIM
	for (k3 = 0; k3<D3; k3++) 	
    	{
		ix2 =  k3*D1;	// index for output 
		ix1 = ix2*D2;	// index for input 

		// Inner LOOP: along dimensions < DIM
		for (k1 = 0; k1<D1; k1++, ix1++, ix2++) 	
		{
		  	s  = 0.0;
		  	N  = 0;
		  	s2 = 0.0;
		  	s4 = 0.0;
		
/*			if (0) //(IS_COMPLEX)
			{
				si = 0.0;
				// LOOP  along dimension DIM
		    		for (k2=0; k2<D2; k2++) 	
				{
					aux = imag(xin1(ix1 + k2*D1));
									
					if (aux==aux)	// test for NaN
					{ 
						si+= aux;
						s2+= aux*aux; 
					}	
	 			}
				xout1i(ix2) = si;
			}
*/			
			if (nargout==4)
			{
				// LOOP  along dimension DIM
		    		for (k2=0; k2<D2; k2++) 	
				{
					aux = real(xin1(ix1 + k2*D1));
									
					if (aux==aux)	// test for NaN
					{ 
						N++; 
						s   += aux; 
						aux *= aux;
						s2  += aux; 
						s4  += aux*aux; 
					}	
	 			}
				xout1(ix2) = s;
				xout2(ix2) = (double) N;
				xout3(ix2) = s2;
				xout4(ix2) = s4;
			}	
			else  if (nargout==3)
			{
				// LOOP  along dimension DIM
		    		for (k2=0; k2<D2; k2++) 	
				{
					aux = real(xin1(ix1 + k2*D1));
									
					if (aux==aux)	// test for NaN
					{ 
						N++; 
						s   += aux; 
						s2  += aux*aux; 
					}	
	 			}
				xout1(ix2) = s;
				xout2(ix2) = (double) N;
				xout3(ix2) = s2;
			}	
			else if (nargout<=2)   // (arg.is_real_type ()) // if (arg.is_complex_type ())
			{
				// LOOP  along dimension DIM
		    		for (k2=0; k2<D2; k2++) 	
				{
					aux =  real(xin1(ix1 + k2*D1));
									
					if (aux==aux)	// test for NaN
					{ 
						N++; 
						s+= aux; 
					}	
	 			}
				xout1(ix2) = s;
				xout2(ix2) = (double) N;
			}
		}
  	}  

/*	if (IS_COMPLEX)
		retval.append(octave_value(xout1,xout1i));
	else
*/
	retval.append(octave_value(xout1));
	retval.append(octave_value(xout2));
	if (nargout>2)
	  	retval.append(octave_value(xout3));
	if (nargout>3)
	  	retval.append(octave_value(xout4));
	
}

File: dependencies/SIFT-private/external/tsa/minimal/sumskipnan.m
function [o,count,SSQ,S4M] = sumskipnan(x,DIM)
% SUMSKIPNAN adds all non-NaN values. 
%
% All NaN's are skipped; NaN's are considered as missing values. 
% SUMSKIPNAN of NaN's only  gives O; and the number of valid elements is return. 
% SUMSKIPNAN is also the elementary function for calculating 
% various statistics (e.g. MEAN, STD, VAR, RMS, MEANSQ, SKEWNESS, 
% KURTOSIS, MOMENT, STATISTIC etc.) from data with missing values.  
% SUMSKIPNAN implements the DIMENSION-argument for data with missing values.
% Also the second output argument return the number of valid elements (not NaNs) 
% 
% Y = sumskipnan(x [,DIM])
% [Y,N,SSQ] = sumskipnan(x [,DIM])
% 
% DIM	dimension
% Y	resulting sum
% N	number of valid (not missing) elements
% SSQ	sum of squares
%
% the function FLAG_NANS_OCCURED() returns whether any value in x
%  is a not-a-number (NaN)
%
% features:
% - can deal with NaN's (missing values)
% - implements dimension argument. 
% - compatible with Matlab and Octave
%
% see also: FLAG_NANS_OCCURED, SUM, NANSUM, MEAN, STD, VAR, RMS, MEANSQ, 
%      SSQ, MOMENT, SKEWNESS, KURTOSIS, SEM


%    This program is free software; you can redistribute it and/or modify
%    it under the terms of the GNU General Public License as published by
%    the Free Software Foundation; either version 2 of the License, or
%    (at your option) any later version.
%
%    This program is distributed in the hope that it will be useful,
%    but WITHOUT ANY WARRANTY; without even the implied warranty of
%    MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
%    GNU General Public License for more details.
%
%    You should have received a copy of the GNU General Public License
%    along with this program; If not, see <http://www.gnu.org/licenses/>.

%	$Id: sumskipnan.m 5599 2009-03-10 12:52:20Z schloegl $
%    	Copyright (C) 2000-2005,2009 by Alois Schloegl <a.schloegl@ieee.org>	
%       This function is part of the NaN-toolbox
%       http://www.dpmi.tu-graz.ac.at/~schloegl/matlab/NaN/


global FLAG_NANS_OCCURED;

if nargin<2,
        DIM = [];
end;

% an efficient implementation in C of the following lines 
% could significantly increase performance 
% only one loop and only one check for isnan is needed
% An MEX-Implementation is available in sumskipnan.cpp
%
% Outline of the algorithm: 
% for { k=1,o=0,count=0; k++; k<N} 
% 	if ~isnan(i(k)) 
% 	{ 	o     += x(k);
%               count += 1;
%		tmp    = x(k)*x(k)
%		o2    += tmp;
%		o3    += tmp.*tmp;
%       }; 

if isempty(DIM),
        DIM = min(find(size(x) > 1));
        if isempty(DIM), DIM = 1; end;
end
if (DIM<1) DIM = 1; end; %% Hack, because min([])=0 for FreeMat v3.5


%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% non-float data 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if ~isa(x,'float') %%%% || ~flag_implicit_skip_nan(), %%% skip always NaN's
	x = double(x); 
	o = sum(x,DIM);
	if nargin>1
		sz = size(x);	
		count = repmat(sz(DIM),sz([1:DIM-1,DIM+1:end]));
		if nargin>2
			x = x.*x; 
			SSQ = sum(x,DIM);
			if nargin>3
				S4M = sum(x.*x,DIM);
			end; 
		end; 
	end; 	
	return; 
end; 	

%% mex and oct files expect double
x = double(x); 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% use Matlab-MEX function when available  
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%%if 0,	
try
	
	%% using sumskipnan_mex.mex

	%% !!! hack: FLAG_NANS_OCCURED is an output argument, reserve memory !!!
	if isempty(FLAG_NANS_OCCURED),
		FLAG_NANS_OCCURED = logical(0);  % default value 
	end;
	if (nargout<3),
		[o,count] = sumskipnan_mex(x,DIM,FLAG_NANS_OCCURED);
		return; 
	elseif (nargout==3),
		[o,count,SSQ] = sumskipnan_mex(x,DIM,FLAG_NANS_OCCURED);
		return; 
	elseif (nargout==4) && isreal(i),
		[o,count,SSQ,S4M] = sumskipnan_mex(x,DIM,FLAG_NANS_OCCURED);
		return; 
	end; 	
end; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% use OCT function when available  
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
%% no performance analysis showing a distinct advantage has been performed.  
if 0 %%if exist('OCTAVE_VERSION','builtin'),

	%% for Octave, there might be an sumskipnan_oct.oct
	try
	if isreal(x)
		%% using sumskipnan_oct.mex
		if (nargout<3),
			[o,count] = sumskipnan_oct(x,DIM);
			FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
			return; 
		elseif (nargout==3),
			[o,count,SSQ] = sumskipnan_oct(x,DIM);
			FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
			return; 
		elseif (nargout==4) && isreal(i),
			[o,count,SSQ,S4M] = sumskipnan_oct(x,DIM);
			FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
			return; 
		end; 	
	else	%% if ~isreal(x)
		if (nargout<3),
			[or,count] = sumskipnan_oct(real(x),DIM);
			[oi,counti] = sumskipnan_oct(imag(x),DIM);
			if any(count(:)-counti(:))
				error('number of NaN differ between real and imag part\n');
			end;
			o = or + i*oi;
			FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
			return; 
		elseif (nargout<4),
			[or,count,SSQ] = sumskipnan_oct(real(x),DIM);
			[oi,counti,SSQi] = sumskipnan_oct(imag(x),DIM);
			if any(count(:)-counti(:))
				error('number of NaN differ between real and imag part\n');
			end;
			o = or + oi;
			SSQ = SSQ+SSQi; 
			FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
			return; 
		end; 	
	end;	%% if 
	end; 	%% try
end; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% fall-back in case no OCT/MEX support is avialable   
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%

%% performance tweak: some tests have shown that x*ones(:,1) is faster than sum(x,2)
FLAG	 = (length(size(x))<3); 
if FLAG, FLAG = DIM; end; 

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% count non-NaN's
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
if nargout>1,
        if FLAG~=2,
                count = sum(x==x,DIM); 
        else
                count = real(x==x)*ones(size(x,2),1);
        end;
	FLAG_NANS_OCCURED = any(count(:)<size(x,DIM));
end;

%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
% replace NaN's with zero 
%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%%
x(x~=x) = 0;	

if FLAG~=2, 
        o = sum(x,DIM);
else 
	o = x*ones(size(x,2),1);
end;

if nargout>2,
        x = real(x).^2 + imag(x).^2;
        if FLAG~=2,
                SSQ = sum(x,DIM);
        else
                SSQ = x*ones(size(x,2),1);
        end;
        if nargout>3,
                S4M = sumskipnan(x.^2,DIM);
        end;
end;

